--- a/arch/arm/kernel/module.c	2019-01-24 21:35:20.188001000 +0800
+++ b/arch/arm/kernel/module.c	2019-01-20 14:40:26.251167688 +0800
@@ -40,11 +40,19 @@
 #ifdef CONFIG_MMU
 void *module_alloc(unsigned long size)
 {
-	void *p = __vmalloc_node_range(size, 1, MODULES_VADDR, MODULES_END,
-				GFP_KERNEL, PAGE_KERNEL_EXEC, 0, NUMA_NO_NODE,
+	unsigned long gfp_flags = GFP_KERNEL;
+	void *p;
+
+	if (IS_ENABLED(CONFIG_ARM_MODULE_PLTS))
+		gfp_flags = gfp_flags | __GFP_NOWARN;
+
+	p = __vmalloc_node_range(size, 1, MODULES_VADDR, MODULES_END,
+				gfp_flags, PAGE_KERNEL_EXEC, 0, NUMA_NO_NODE,
 				__builtin_return_address(0));
+
 	if (!IS_ENABLED(CONFIG_ARM_MODULE_PLTS) || p)
 		return p;
+
 	return __vmalloc_node_range(size, 1,  VMALLOC_START, VMALLOC_END,
 				GFP_KERNEL, PAGE_KERNEL_EXEC, 0, NUMA_NO_NODE,
 				__builtin_return_address(0));
